addons/arsenal/functions/fnc_fillLoadoutsList.sqf

```diff
@@ -1,42 +1,60 @@
 #include "..\script_component.hpp"
 #include "..\defines.hpp"
 /*
- * Author: Alganthe, johnb43
+ * Author: Alganthe, johnb43, LinkIsGrim
  * Fill loadouts list.
  *
  * Arguments:
  * 0: Loadouts display <DISPLAY>
  * 1: Tab control <CONTROL>
+ * 2: Current frame filling loadouts list <NUMBER> (default: 0)
+ * 3: Frames necessary to fill loadouts list <NUMBER> (default: -1)
  *
  * Return Value:
  * None
  *
  * Public: No
 */
 
-params ["_display", "_control"];
+#define LOADOUTS_PER_FRAME 10
 
-(_display displayCtrl IDC_textEditBox) ctrlSetText "";
+params ["_display", "_control", ["_currentFrame", 0], ["_framesToFill", -1]];
+
+if (isNull _display) exitWith {
+    TRACE_2("display closed, aborting",_currentFrame,_framesToFill);
+};
 
 private _contentPanelCtrl = _display displayCtrl IDC_contentPanel;
+if (_currentFrame == 0) then {
+    (_display displayCtrl IDC_textEditBox) ctrlSetText "";
 
-// Force a "refresh" animation of the panel
-_contentPanelCtrl ctrlSetFade 1;
-_contentPanelCtrl ctrlCommit 0;
-_contentPanelCtrl ctrlSetFade 0;
-_contentPanelCtrl ctrlCommit FADE_DELAY;
+    // Force a "refresh" animation of the panel
+    _contentPanelCtrl ctrlSetFade 1;
+    _contentPanelCtrl ctrlCommit 0;
+    _contentPanelCtrl ctrlSetFade 0;
+    _contentPanelCtrl ctrlCommit FADE_DELAY;
 
-_contentPanelCtrl lnbSetCurSelRow -1;
-lnbClear _contentPanelCtrl;
+    _contentPanelCtrl lnbSetCurSelRow -1;
+    lnbClear _contentPanelCtrl;
+};
 
 private _sharedLoadoutsVars = GVAR(sharedLoadoutsNamespace) getVariable QGVAR(sharedLoadoutsVars);
-private _cfgWeapons = configFile >> "CfgWeapons";
+private _cfgWeapons = configFile >> "CfgWeapons"; // Used by ADD_LOADOUTS_LIST_PICTURES macro, do not remove
 private _newRow = -1;
 
 if (GVAR(currentLoadoutsTab) != IDC_buttonSharedLoadouts) then {
     private _loadoutNameAndTab = "";
     private _loadoutCachedInfo = "";
     private _sharingEnabled = GVAR(allowSharedLoadouts) && {isMultiplayer};
+    private _loadouts = [
+        profileNamespace getVariable [QGVAR(saved_loadouts), []],
+        GVAR(defaultLoadoutsList)
+    ] select (ctrlIDC _control == IDC_buttonDefaultLoadouts);
```

Comment:
Isn't there a chance that these variables could change over the course of loading?

Reply:
Contents could change, menu would need to be refilled for changes to take effect

Reply:
My question is: Can it break due to that?

Reply:
loadouts with index divisible by `LOADOUTS_PER_FRAME` + number of deleted loadouts (so the index 10 until 10 + n, 20->20 +n, 30->30 +n... loadouts, where n is how many loadouts with lower indexes were removed from the array) could potentially be skipped if a loadout from an earlier batch was removed from the loadout array before next frame.

Loadouts that have their contents altered would keep their contents at whatever frame they were processed (loadout L1 is added at frame 0, get its contents changed before frame 1, list is filled during frame 1, if a player tries to load L1 they'll load the content it had during frame 0.

Due to the syntax of select used the only hard break is nil-ing the loadout array entirely.

Reply:
What about multiple function calls, one every frame?

Reply:
only way for that to happen is for the display to be opened & closed multiple times, at which point the loop will abort on a null display